Lib/asyncio/locks.py

```diff
@@ -412,3 +412,209 @@ def release(self):
+        self._state = 0       # 0 filling, 1, draining, -1 resetting, -2 broken
+        self._count = 0       # count tasks in Barrier
+        self._count_block = 0 # count blocking tasks
+
+    def __repr__(self):
+        res = super().__repr__()
+        extra = 'locked' if self._cond.locked() else 'unlocked'
+        if self._count:
+            extra += f', wait:{self._count}/{self._parties}'
+        if self._count_block:
+            extra += f', block:{self._count_block}/{self._parties}'
+        extra += f', state:{self._state}'
+        return f'<{res[1:-1]} [{extra}]>'
+
+    async def __aenter__(self):
+        """ wait for the barrier reaches the parties number
+        when start draining release and return index of waited task
+        """
+        return await self.wait()
+
+    async def __aexit__(self, *args):
+        pass
+
+    async def wait(self):
+        """Wait for the barrier.
+        When the specified number of tasks have started waiting, they are all
+        simultaneously awoken. If an 'action' was provided for the barrier, the
+        last task calling this method will have executed that callback prior to
+        returning.
+        Returns an unique and individual index number from 0 to 'parties-1'.
+        # """
+        async with self._cond:
+            await self._block() # Block while the barrier drains or resets.
+            try:
+                index = self._count
+                self._count += 1
+                if index + 1 == self._parties:
+                    # We release the barrier
+                    await self._release()
+                else:
+                    await self._wait()
+                return index
+            finally:
+                self._count -= 1
+                # Wake up any tasks waiting for barrier to drain.
+                self._exit()
+
+    async def _block(self):
+        """Block until the barrier is ready for us,
+        or raise an exception if it is broken.
+        """
+        self._count_block += 1
+        # It is draining or resetting, wait until done
+        # unless a CancelledError occurs
+        try:
+            await self._cond.wait_for(lambda: not (self._draining() or
+                                                    self._resetting()))
+        except exceptions.CancelledError:
+            self._count_block -= 1
+            raise
+        self._count_block -= 1
+
+        # see if the barrier is in a broken state
+        if self.broken:
+            raise BrokenBarrierError
+
+    # Optionally run the 'action' and release the tasks waiting
+    # in the barrier.
+    async def _release(self):
+        """Optionally run the 'action' and release the tasks waiting
+        in the barrier.
+        """
+        try:
+            if self._action:
+                await self._action()
+            # enter draining state
+            self._set_draining()
+            self._cond.notify_all()
+        except:
```

Comment:
We shouldn't have any bare `except`. Replace with `except Exception`. Moreover, if this is to catch errors in `_action`, the `await self._action` should be surrounded with `try..except` and nothing else.

Reply:
done